File: DataViz/Utilities/SensorSmoothing.swift
import Foundation

/// A simple low-pass filter for smoothing noisy sensor data.
///
/// The smoothing factor `alpha` controls responsiveness:
/// - Values closer to `1.0` track the input closely (less smoothing).
/// - Values closer to `0.0` respond slowly to changes (more smoothing).
struct LowPassFilter {
    let alpha: Double
    private var previousValue: Double?

    /// Creates a low-pass filter with the given smoothing factor.
    /// - Parameter alpha: A value in the range `0.0...1.0`. Defaults to `0.2`.
    init(alpha: Double = 0.2) {
        precondition((0.0...1.0).contains(alpha), "alpha must be between 0.0 and 1.0")
        self.alpha = alpha
    }

    /// Applies the filter to a new input value and returns the smoothed result.
    ///
    /// On the first call (or after a reset), the input value is returned unchanged.
    /// Subsequent calls return `alpha * newValue + (1 - alpha) * previousValue`.
    mutating func apply(_ newValue: Double) -> Double {
        guard let previous = previousValue else {
            previousValue = newValue
            return newValue
        }
        let filtered = alpha * newValue + (1.0 - alpha) * previous
        previousValue = filtered
        return filtered
    }

    /// Resets the filter, discarding any accumulated state.
    mutating func reset() {
        previousValue = nil
    }
}

/// Returns `0.0` if the absolute value of the input is below the threshold;
/// otherwise returns the input unchanged.
///
/// Useful for ignoring small sensor fluctuations that represent noise rather
/// than intentional movement.
/// - Parameters:
///   - value: The raw sensor value.
///   - threshold: The minimum absolute value to pass through.
/// - Returns: The original value, or `0.0` if within the dead zone.
func deadZone(_ value: Double, threshold: Double) -> Double {
    abs(value) < threshold ? 0.0 : value
}

File: web/js/sensor-smoothing.js
/**
 * sensor-smoothing.js
 * Low-pass filter and dead-zone utilities for smoothing noisy sensor input.
 */

/**
 * First-order IIR (infinite impulse response) low-pass filter.
 *
 * Each call to `apply()` blends the new sample with the previous output:
 *   output = alpha * newValue + (1 - alpha) * previousOutput
 *
 * A smaller alpha produces heavier smoothing (slower response).
 */
export class LowPassFilter {
    /**
     * @param {number} [alpha=0.2] - Smoothing factor in the range (0, 1].
     *   Values closer to 0 give heavier smoothing; 1 means no smoothing.
     */
    constructor(alpha = 0.2) {
        this.alpha = alpha;
        this._previous = null;
    }

    /**
     * Feed a new raw sample and return the filtered value.
     * @param {number} newValue - The latest sensor reading.
     * @returns {number} Smoothed value.
     */
    apply(newValue) {
        if (this._previous === null) {
            this._previous = newValue;
            return newValue;
        }
        this._previous = this.alpha * newValue + (1 - this.alpha) * this._previous;
        return this._previous;
    }

    /** Reset the filter so the next sample is treated as the first. */
    reset() {
        this._previous = null;
    }
}

/**
 * Dead-zone function: returns 0 when the absolute value of the input
 * is below the given threshold, otherwise returns the original value.
 *
 * Useful for ignoring sensor noise or tiny joystick drift.
 *
 * @param {number} value - Raw input value.
 * @param {number} threshold - Values with |value| < threshold are zeroed.
 * @returns {number} 0 if within the dead zone, otherwise `value`.
 */
export function deadZone(value, threshold) {
    return Math.abs(value) < threshold ? 0 : value;
}
